bug 1436067 - Abstract screen.less into smaller modules

andymikulski opened the pull request:
Fixes [bug #1436067](https://bugzilla.mozilla.org/show_bug.cgi?id=1436067) - abstract `screen.less` file into smaller modules.

This PR should be generally straight forward, as most of it is simply moving code into separate files, with as little modification to the actual content of the code as possible. 

To review, I suggest examining:
  - The file tree changes (outlined below)
   - The LESS global variable addition to `/crashstats/settings/base.py`
   - The LESS files for topcrashers etc to see how imports work (spoiler, like this: `@import "@{base-path}/mixins.less";`)

>Things of note:
>- `base-path` currently points to a directory beginning with `/tmp`, so I'm not sure if it's something that will work properly in production. I found the existing path via an error message from the webapp.
>- `screen.less` has been removed, and `base.less` has taken its place. If this is against django conventions or something, I can keep the file.
>- I am unsure if the elements for the selectors left in the bottom of `base.less` actually exist on the site, but I plan to deal with that in another PR.

I'm certainly open to feedback on every aspect of this, so please let me know if you have any thoughts on anything!

---

<details>
<summary>Old File Structure</summary>

```
/socorro/webapp-django/crashstats/crashstats/static/crashstats/css
├── accordion.less
├── base.less
├── crashes_per_day.less
├── crontabber_state.css
├── fonts.less
├── forms.less
├── lib
│   ├── images/
│   ├── jquery-ui.structure.css
│   └── jquery-ui.theme.css
├── metricsgraphics_custom.css
├── mixins.less
├── report_index.css
├── report_pending.less
├── reset.less
├── screen.less
├── tables.less
├── tree.css
└── variables.less
```

</details>

<details>
<summary>New File Structure</summary>

```
/socorro/webapp-django/crashstats/crashstats/static/crashstats/css
├── base
│   ├── fonts.less
│   ├── forms.less
│   ├── layout.less
│   ├── mixins.less
│   ├── reset.less
│   ├── screen.less
│   ├── tables.less
│   ├── typography.less
│   └── variables.less
├── base.less
├── components
│   ├── accordion.less
│   ├── admin.less
│   ├── code.less
│   ├── footer.less
│   ├── message.less
│   ├── nav.less
│   ├── pagination.less
│   ├── panel.less
│   ├── simplebox.less
│   ├── sumo_link.less
│   ├── table_sorter.less
│   └── tree.less
├── lib
│   ├── images/
│   ├── jquery-ui.structure.css
│   ├── jquery-ui.theme.css
│   └── metricsgraphics_custom.css
└── pages
    ├── crashes_per_day.less
    ├── crontabber_state.css
    ├── report_index.less
    └── report_pending.less
```

</details>

andymikulski commented:
@Osmose , I can't assign you via GitHub, but this is ready for your review!

andymikulski commented on webapp-django/crashstats/crashstats/static/crashstats/css/base.less:
```diff
@@ -1,5 +1,254 @@
+@import "components/nav.less";
+@import "components/footer.less";
+
+/* Common components */
```
Note: the dream is to eventually move these component `import`s to the pages that actually require them (where possible). For now, they are all included here as a temporary measure.

andymikulski commented on webapp-django/crashstats/crashstats/static/crashstats/css/base.less:
```diff
@@ -1,5 +1,254 @@
+
+
+
+/* --- */
```
Below this line are the elements I could not verify exist on the site, and will eventually be removed in a later PR after some thorough site crawling/testing to ensure they really don't exist anywhere.

andymikulski commented:
Whoops, CI is failing due to a bad import. I will correct, strange I didn't catch that before opening..

Osmose commented on webapp-django/crashstats/settings/base.py:
```diff
@@ -447,6 +447,7 @@ def path(*dirs):
         'LESS_BINARY',
         path('node_modules/.bin/lessc')
     ),
+    'LESS_ARGUMENTS': '--global-var="base-path=\'/tmp/crashstats-static/crashstats/css/base\'"',
```
I get the following error when running `make dockerbuild` to build a new base image:

```
pipeline.exceptions.CompilerError: ['/webapp-frontend-deps/node_modules/.bin/lessc', "--global-var=base-path='/tmp/crashstats-static/crashstats/css/base'", u'/app/webapp-django/static/home/css/home.less'] exit code 1
FileError: '/tmp/crashstats-static/crashstats/css/base/variables.less' wasn't found. Tried - /tmp/crashstats-static/crashstats/css/base/variables.less,/app/webapp-django/static/home/css/tmp/crashstats-static/crashstats/css/base/variables.less in /app/webapp-django/static/home/css/home.less on line 2, column 1:
1 @base-path: '/tmp/crashstats-static/crashstats/css/base';
2 @import "@{base-path}/variables.less";
```

I suspect it has to do with the path being passed here. I believe `/tmp/crashstats-static` only exists during local development and not during the build process. The [note on the webapp docs](http://socorro.readthedocs.io/en/latest/components/webapp.html) mentions that normally these files are in `/app/webapp-django/static`, maybe that would work instead?

andymikulski replied:
Thanks for pointing this out, I knew there was something wrong with the `/tmp` business.

I've changed this so it checks if the site is running in dev mode, and switches between the two directories appropriately. This seems to satisfy the build and it works locally, hopefully this approach is acceptable!

andymikulski commented on webapp-django/crashstats/crashstats/static/crashstats/css/base.less:
```diff
@@ -1,5 +1,254 @@
-@import "forms.less";
-@import "tables.less";
+/*
+Copyright (c) 2009, Yahoo! Inc. All rights reserved.
```
I figured it was a safe bet to maintain the YUI license until I've at least put some effort towards ensuring all traces of YUI have been removed. If you think it's safe to remove at this point, I'm happy to do so!

Osmose replied:
That's fair. My worry is that a copyright notice that only covers parts of the file, and doesn't signify which parts, is not-so-great either.

I don't really have a good solution here for the interim. You know what's best.

andymikulski commented:
There were some issues around a handful of LESS files due to their use of relative image paths. I've corrected this and introduced `image-path`, a LESS variable to eliminate the need of using `../../../` anywhere we want to include an image in styles.

willkg commented on webapp-django/crashstats/settings/base.py:
```diff
@@ -439,6 +439,7 @@ def path(*dirs):
 
 STATICFILES_STORAGE = 'pipeline.storage.PipelineCachedStorage'
 
+LESS_PATH = '/tmp/crashstats-static' if LOCAL_DEV_ENV else '/app/webapp-django/static'
```
Instead of hard-coding it, you can use the `STATIC_ROOT` variable. So you can do:

```
'LESS_ARGUMENTS': '--global-var="base-path=\'' + STATIC_ROOT + '/crashstats/css/base\'"',
```

Pretty sure that should work.

In the local dev environment, it's set to `/tmp/crashstats-static/` and in all other environments, it's set to `/app/webapp-django/static` (not sure if that has a `/` at the end or not).

andymikulski replied:
Ah, gotcha. For some reason, I thought the STATIC_ROOT pointed somewhere else entirely. I will correct!

andymikulski commented:
Travis failed due to connection timeouts, and Circle has errors on `/topcrashers`. Any chance these two issues are related? As far as I can tell, what I've touched is not responsible, but perhaps I'm wrong.

Osmose commented:
Just an FYI I'm mostly out for the rest of the day and on PTO tomorrow so I may not get back to this until Thursday.
